BazelExtensions/source_output_file_map_aspect.bzl

```diff
@@ -0,0 +1,78 @@
+"""
+Aspect to map source to output files (i.e. .o files generated during compilation).
+
+The mappings are generated as JSON files per target.
+"""
+
+SourceOutputFileMapInfo = provider(
+    doc = "...",
+    fields = {
+        "mapping": "Dictionary where keys are source file paths and values are at the respective .o file under bazel-out",
+    },
+)
+
+def _source_output_file_map_aspect_impl(target, ctx):
+    source_output_file_map = ctx.actions.declare_file("{}_source_output_file_map.json".format(target.label.name))
+
+    mapping = {}
+    objc_srcs = []
+    objc_objects = []
+
+    # List of source files to be mapped to output files
+    if hasattr(ctx.rule.attr, "srcs"):
+        objc_srcs = [
+            f
+            for source_file in ctx.rule.attr.srcs
+            for f in source_file.files.to_list()
+            # Handling objc only for now
+            if f.path.endswith(".m")
+            # TODO: handle swift
+        ]
+
+    # Get compilation outputs if present
+    if OutputGroupInfo in target:
+        if hasattr(target[OutputGroupInfo], "compilation_outputs"):
+            objc_objects.extend(target[OutputGroupInfo].compilation_outputs.to_list())
+
+    # Map source to output file
+    if len(objc_srcs):
+        if len(objc_srcs) != len(objc_objects):
+            fail("[ERROR] Unexpected number of object files")
+        for src in objc_srcs:
+            basename_without_ext = src.basename.replace(".%s" % src.extension, "")
+            obj = [o for o in objc_objects if "%s.o" % basename_without_ext == o.basename]
+            if len(obj) != 1:
+                fail("Failed to find single object file for source %s. Found: %s" % (src, obj))
+
+            obj = obj[0]
+            mapping["/{}".format(src.path)] = obj.path
+
+    # Collect info from deps
+    deps = getattr(ctx.rule.attr, "deps", [])
+    transitive_jsons = []
+    for dep in deps:
+        # Collect mappings from deps
+        if SourceOutputFileMapInfo in dep:
+            for k, v in dep[SourceOutputFileMapInfo].mapping.items():
+                mapping[k] = v
+        # Collect generated JSON files from deps
+        if OutputGroupInfo in dep:
+            if hasattr(dep[OutputGroupInfo], "source_output_file_map"):
+                transitive_jsons.extend(dep[OutputGroupInfo].source_output_file_map.to_list())
+
+    # Writes JSON
+    ctx.actions.write(source_output_file_map, json.encode(mapping))
+
+    return [
+        OutputGroupInfo(
+            source_output_file_map = depset([source_output_file_map] + transitive_jsons),
+        ),
+        SourceOutputFileMapInfo(
+            mapping = mapping
+        ),
+    ]
+
+source_output_file_map_aspect = aspect(
```

Comment:
A couple suggestion:
- can we align this with the existing aspect to prevent another traversal
- do you want to tack this onto the `xcodeproj` rule `deps` attr - so that these files are generated at _project_ generation time or not?

Reply:
>can we align this with the existing aspect to prevent another traversal

Done

>do you want to tack this onto the xcodeproj rule deps attr - so that these files are generated at project generation time or not?

That would be helpful, to have the aspect output groups ready on Xcode launch. I played with this for a bit but hit issues due to not being able to pre-generate in a transition mappings containing the correct `bazel-out` paths. Essentially the `ST-hash` was different on "aspect + transition" vs "XCHammer build". I believe this is a bug but haven't spent more time digging, since everything works once at least a build in Xcode happens I think we can punt this for now.

TL;DR from my investigation: building with cmd line options `--apple_platform_type=ios --cpu=ios_sim_arm64` is **not** the same as using a transition that transitions to the *exactly same* flags. I believe this is related to this: https://github.com/bazelbuild/bazel/issues/12171